spec/models/results/web_response_parser_spec.rb

```diff
@@ -0,0 +1,178 @@
+    let(:cat) { form.c[1].option_set.sorted_children[1].id }
+    let(:answers) do
+      {
+        "0" => web_answer_hash(form.c[0].id, value: "A"),
+        "1" => web_answer_hash(form.c[1].id,
+          choices_attributes: {
+            "0" => {option_node_id: dog, checked: "1"},
+            "1" => {option_node_id: cat, checked: "1"}
+          }),
+        "2" => web_answer_hash(form.c[2].id, value: "D")
+      }
+    end
+
+    it "builds tree with answer set" do
+      tree = Results::WebResponseParser.new.parse(input)
+      expect_root(tree, form)
+      expect_children(tree, %w[Answer Answer Answer], form.c.map(&:id), ["A", "Cat;Dog", "D"])
+    end
+  end
+
+  context "response with a group" do
+    let(:question_types) { ["text", %w[text text], "text"] }
+    let(:data) do
+      {root: web_answer_group_hash(form.root_group.id,
+        "0" => web_answer_hash(form.c[0].id, value: "A"),
+        "1" => web_answer_group_hash(form.c[1].id,
+          "0" => web_answer_hash(form.c[1].c[0].id, value: "B"),
+          "1" => web_answer_hash(form.c[1].c[1].id, value: "C")),
+        "2" => web_answer_hash(form.c[2].id, value: "D"))}
+    end
+
+    it "should produce the correct tree" do
+      tree = Results::WebResponseParser.new.parse(input)
+      expect_root(tree, form)
+      expect_children(tree, %w[Answer AnswerGroup Answer], form.c.map(&:id), ["A", nil, "D"])
+      expect_children(tree.c[1], %w[Answer Answer], form.c[1].c.map(&:id), %w[B C])
+    end
+  end
+
+  context "response with an answer group set" do
+    let(:question_types) { ["text", {repeating: {items: %w[text text]}}] }
+    let(:data) do
+      {root: web_answer_group_hash(form.root_group.id,
+        "0" => web_answer_hash(form.c[0].id, value: "A"),
+        "1" => {
+          id: "",
+          type: "AnswerGroupSet",
+          questioning_id: form.c[1].id,
+          relevant: "true",
+          children: {
+            "0" => web_answer_group_hash(form.c[1].id, answers1),
+            "1" => web_answer_group_hash(form.c[1].id, answers2)
+          }
+        })}
+    end
+    let(:answers1) do
+      {
+        "0" => web_answer_hash(form.c[1].c[0].id, value: "B"),
+        "1" => web_answer_hash(form.c[1].c[1].id, value: "C")
+      }
+    end
+    let(:answers2) do
+      {
+        "0" => web_answer_hash(form.c[1].c[0].id, value: "D"),
+        "1" => web_answer_hash(form.c[1].c[1].id, value: "E")
+      }
+    end
+
+    it "builds tree with answer group set" do
+      tree = Results::WebResponseParser.new.parse(input)
+      expect_root(tree, form)
+      expect_children(tree, %w[Answer AnswerGroupSet], form.c.map(&:id), ["A", nil])
+      expect_children(tree.c[1], %w[AnswerGroup AnswerGroup], [form.c[1].id, form.c[1].id])
+      expect_children(tree.c[1].c[0], %w[Answer Answer], form.c[1].c.map(&:id), %w[B C])
+      expect_children(tree.c[1].c[1], %w[Answer Answer], form.c[1].c.map(&:id), %w[D E])
+    end
+  end
```

Comment:
Are you pretty confident that multiple levels of nesting will work properly here? Might it be a good idea to add a quick spec for that? Am I right that you've only tested a single level of nesting so far in here?

Reply:
will add a nested spec